### PR TITLE
Improve and harden router deletion on agent

### PR DESCRIPTION
**Pull request by sebageek**
This is not yet a fix for the agent unscheduling problem, but this is at least half the way there.

### CODE REVIEW

File: asr1k_neutron_l3/common/cache_utils.py [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+    key = _gen_cache_key("deleted-router", host, router_id)
+    cache.set(key, data)
+
+
+def get_deleted_router(host, router_id):
+    cache = get_cache()
+    if not cache:
+        return
+
+    key = _gen_cache_key("deleted-router", host, router_id)
```
Comment (joker-at-work) [inline asr1k_neutron_l3/common/cache_utils.py]: Would probably make sense to move this prefix into a variable as it's used in two places.

File: asr1k_neutron_l3/plugins/l3/rpc/rpc_api.py [hunk 2]
```diff
@@ -55,11 +56,12 @@ def update_router_status(self, context, **kwargs):
...
-    def get_deleted_routers(self, context, **kwargs):
-        host = kwargs.get('host')
-        router_ids = kwargs.get('router_ids')
-
-        return self.l3plugin.get_sync_data(context, router_ids=router_ids, active=None, host=host)
+    def get_deleted_router(self, context, host, router_id):
+        router = cache_utils.get_deleted_router(host, router_id)
+        if not router:
+            LOG.debug("No entry for router %s on host %s in deleted router cache", router_id, host)
+            return None
```
Comment (joker-at-work) [inline asr1k_neutron_l3/plugins/l3/rpc/rpc_api.py]: This `None` would not be necessary.

**Review by joker-at-work — COMMENTED** [review]
In general, do we expect any problems with serializing the data of the router into the cache? Are there any objects, that aren't used too often, that might fail the serialization?

**Comment by sebageek** [comment]
From what I've seen the output of `get_sync_data()` only contains dicts, lists, strings and numbers, so serializing should not be a problem. I didn't follow all code paths though. On the other hand this datastructure at least gets serialized for transport through the rpc framework from neutron to the asr1k agent.

**Review by joker-at-work — APPROVED** [review]
If it's going over the wire, then neutron probably removed all the problematic objects or has compatibility shims for serialization. :+1:

**Pull request merged**